### PR TITLE
Selector functions

### PR DESCRIPTION
**Pull request by onedayitwillmake**
**Please note: The code in this pull request was written by myself, and is does not represent my employer.** 

*updated 4/08/15*
*updated 4/09/15*

I rebased the PR https://github.com/sass/libsass/pull/1019 and removed Xcode references.

I have 8 of the 8 built, I started from the bottom for some reason.
Bold means they are in and passing the test. 

I'm still actively working on this - however being my second PR and new to the library I wanted to make sure anyone could catch any silly things I was doing.

- [x] selector-nest($selectors...)
- [x] selector-append($selectors...)
- [x] selector-extend($selector, $extendee, $extender)
- [x] selector-replace($selector, $original, $replacement)
- [x] selector-parse($selector)
- [x] simple-selectors($selector)
- [x] is-superselector($super, $sub)
- [x] selector-unify($selector1, $selector2)

Some notes:
* Added stub functions
* Passing 'Contextualize' to `Native_Function signature in BUILT_IN(name) Expression*
* Added `ARGSELL` to compliment `ARG`, `ARGR`, `ARGM` since all the selector functions rely on that so heavily 
* Had to do some hairy things to get `Selector_List::selector_unify` and `Selector_List::selector_unify` to work
* Added `public static` methods version of Extend's subweave, *for now* I named it StaticSubweave just to make sure I don't step on any toes, will remove floating one if that seems fine to everyone
* Added a naiveTrim function to Node that removes any duplicate selectors (this is for `Complex_Selector::unify_with` was returning correct results but the final shared selector would sometimes appear twice, and blindly adding everything else in the collection (I could not get node's existing `trim` implementation to work - I Also could not find any references to it be used.
* Added a `public static` version of Extend's extendSelectorList
* Added param `isReplace` to `Extend::extendSelectorList` - which does not add the Selector if an extension takes place. 
* Added `Vectorized<T>::first` to compliment  `Vectorized<T>::back` in order to make `selector_append` code more clear,

### CODE REVIEW

**Comment by mgreter** [comment]
Hey @onedayitwillmake great work so far :+1: 

I guess it is already passing quite a few todo tests? I also expect a few other extend bugs to be solved by this work, since it should fix/improve the `is_superselector` function. Maybe you have such a list? Otherwise I could add it since I have pretty much everything in place to generate it automatically with perl-libsass (IMO it should also be possible via sassc, but I never use it)!

Will definitely give it a thorough review after we released 3.2 :fireworks:

**Comment by mgreter** [comment]
Gave it a whirl and these tests pass now on my side :smiley: 

- selector-functions/extend
- selector-functions/parse
- selector-functions/replace
- selector-functions/extend/nested
- selector-functions/is_superselector/any_is_not_superselector_of_different_prefix
- selector-functions/is_superselector/complex_superselector
- selector-functions/is_superselector/compound_superselector
- selector-functions/is_superselector/current_is_superselector_with_identical_innards
- selector-functions/is_superselector/current_is_superselector_with_subselector_innards
- selector-functions/is_superselector/descendant_is_superselector_of_child
- selector-functions/is_superselector/following_sibling_is_superselector_of_next_sibling
- selector-functions/is_superselector/leading_combinator
- selector-functions/is_superselector/matches_is_not_superselector_of_any
- selector-functions/is_superselector/matching_combinator
- selector-functions/is_superselector/not_is_not_superselector_of_non_unique_selectors
- selector-functions/is_superselector/reflexivity
- selector-functions/is_superselector/selector_list_subset
- selector-functions/is_superselector/trailing_combinator

**Comment by onedayitwillmake** [comment]
@mgreter Would be great if you could generate it! 

@mgreter @xzyfer Re: $selector-append 
I have the function working with functionally correct output, however in ruby-sass the output is ordered alphabetically. 

To give an example:
```
input: selector-append(".a", ".b, .c", ".de, .df, foo");
// ruby-sass
.a.b.de, .a.b.df, .a.bfoo, .a.c.de, .a.c.df, .a.cfoo;
// libsass 
.a.b.de, .a.c.de, .a.b.df, .a.c.df, .a.bfoo, .a.cfoo;
```

I could sort all the Complex_Selectors in the Selector_List, after the operation - but I'm not sure if that would be the right thing to do. Any advice?

**Comment by onedayitwillmake** [comment]
@mgreter @xzyfer Alright! Implemented `selector-nest` to make the full set

**Comment by onedayitwillmake** [comment]
@mgreter @xzyfer Do you guys have any advice on fixing the build errors in some of the travis-ci builds? It looks like it's failing on the `issue_1075`, in only some of the builds

**Comment by xzyfer** [comment]
@onedayitwillmake if you apply this change you should be good

screen shot 2015-04-10 at 1 43 08 pm

**Comment by onedayitwillmake** [comment]
@xzyfer @mgreter Just curious if there was any movement on this, I implemented the `.travis.yml` modifications, should I just rebase?

**Comment by xzyfer** [comment]
@onedayitwillmake yes please rebase on master and resolve any conflicts. 

It also looks like this branch has had a lot of merges with master which means it cannot be merged in it's current state. A rebase may help push out the unwanted commits, otherwise we'll need to find another way to clean up this branch.

We rely heavily on git forensics to track down bug and regressions. This requires a clean commit history.

**Comment by onedayitwillmake** [comment]
@xzyfer Gotcha, makes sense.
Besides the one on the bottom, which other ones do you see have merges with master?

**Comment by xzyfer** [comment]
I can see commits from myself and @mgreter which don't belong here. These were presumably introduced in a merge.

screen shot 2015-04-14 at 10 37 52 am
screen shot 2015-04-14 at 10 38 03 am

**Comment by xzyfer** [comment]
@onedayitwillmake have you had a chance to clean up this PR? You can probably squash this down to a few commits.

**Comment by onedayitwillmake** [comment]
@xzyfer I will give it another go tomorrow morning, you reckon I should just squash it to a few key commits and rebase from there?

**Comment by xzyfer** [comment]
@onedayitwillmake yes I do. However I expect you will into issues due to the merge commits.

**Comment by onedayitwillmake** [comment]
@xzyfer Looks like something else broke on this one, saying I have a duplicate function however I'm not seeing that on my end. Maybe I'm just bad at pull-request, what do you suggest

File: ast.cpp [hunk 1]
```diff
@@ -490,6 +494,215 @@ namespace Sass {
...
+      if(!any) {
+        return false;
+      }
+    }
+    return true;
+  }
+  
+
+  // For every selector in RHS, see if we have /any/ selectors which are a super-selector of it
+  bool Selector_List::is_superselector_of(Sass::Selector_List *rhs) {
```
Comment (xzyfer) [inline ast.cpp]: I believe this is cause of the CI error. See line 501 above.

**Comment by xzyfer** [comment]
Impressive git work to fix this branch.

Please remove the debug code.

**Comment by onedayitwillmake** [comment]
@xzyfer Thanks for heads up, I removed them

**Comment by onedayitwillmake** [comment]
@mgreter Any movement on this one?

File: ast.cpp [hunk 2]
```diff
@@ -492,6 +496,131 @@ namespace Sass {
...
+      
+      if (!pIter->head() || pIter->tail()) {
+        error("nested selectors may not be extended", c->pstate());
+      }
+      
+      compound_sel->is_optional(extendee->is_optional());
+      
+      for (size_t i = 0, L = extender->length(); i < L; ++i) {
+        // let's test this out
+        cerr << "REGISTERING EXTENSION REQUEST: " << (*extender)[i]->perform(&to_string) << " <- " << compound_sel->perform(&to_string) << endl;
```
Comment (xzyfer) [inline ast.cpp]: debug code?

File: ast.hpp [hunk 3]
```diff
@@ -633,9 +634,11 @@ namespace Sass {
   // by a type tag.
   /////////////////////////////////////////////////////////////////////////////
   struct Backtrace;
+  class Contextualize;
   typedef Environment<AST_Node*> Env;
   typedef const char* Signature;
-  typedef Expression* (*Native_Function)(Env&, Env&, Context&, Signature, ParserState, Backtrace*);
+  typedef Expression* (*Native_Function)(Env&, Env&, Context&, Signature, ParserState, Contextualize*, Backtrace*);
+//  typedef Expression* (*Native_Function)(Env&, Env&, Context&, Signature, ParserState, Contextualize*, Backtrace*);
```
Comment (xzyfer) [inline ast.hpp]: remove please

File: eval.cpp [hunk 4]
```diff
@@ -624,7 +624,7 @@ namespace Sass {
       } else if (sass_value_get_tag(c_val) == SASS_WARNING) {
         error("warning in C function " + c->name() + ": " + sass_warning_get_message(c_val), c->pstate(), backtrace);
       }
-      result = cval_to_astnode(c_val, ctx, backtrace, c->pstate());
+      result = cval_to_astnode(c_val, ctx, backtrace, c->pstate ());
```
Comment (xzyfer) [inline eval.cpp]: undo this change please

File: expand.cpp [hunk 5]
```diff
@@ -201,6 +201,9 @@ namespace Sass {
     String* old_p = d->property();
     String* new_p = static_cast<String*>(old_p->perform(eval->with(env, backtrace)));
     Selector* p = selector_stack.size() <= 1 ? 0 : selector_stack.back();
+    
+    //    Contextualize_Eval* contextual = contextualize_eval->with(selector_stack.back(), env, backtrace);
```
Comment (xzyfer) [inline expand.cpp]: remove please

File: expand.cpp [hunk 6]
```diff
@@ -493,7 +496,7 @@ namespace Sass {
       // { target_vec.push_back((*compound_sel)[i]->perform(&to_string)); }
       for (size_t i = 0, L = extender->length(); i < L; ++i) {
         // let's test this out
-        // cerr << "REGISTERING EXTENSION REQUEST: " << (*extender)[i]->perform(&to_string) << " <- " << compound_sel->perform(&to_string) << endl;
+         cerr << "REGISTERING EXTENSION REQUEST: " << (*extender)[i]->perform(&to_string) << " <- " << compound_sel->perform(&to_string) << endl;
```
Comment (xzyfer) [inline expand.cpp]: debug code??

File: extend.cpp [hunk 7]
```diff
@@ -1294,6 +1296,168 @@ namespace Sass {
...
+      
+      Node& child = *pathsIter;
+      child = flatten(child, ctx);
+    }
+    
+    DEBUG_PRINTLN(SUBWEAVE, "FLATTENED: " << pathsResult)
+    
+    
+    /*
+     TODO: implement
```
Comment (xzyfer) [inline extend.cpp]: still a todo? if not please remove

File: extend.cpp [hunk 8]
```diff
@@ -1517,7 +1679,7 @@ namespace Sass {
       Complex_Selector& seq = groupedPair.first;
       vector<ExtensionPair>& group = groupedPair.second;
 
-//      DEBUG_EXEC(EXTEND_COMPOUND, printComplexSelector(&seq, "SEQ: "))
```
Comment (xzyfer) [inline extend.cpp]: please re-comment out these.

File: functions.cpp [hunk 9]
```diff
@@ -1566,6 +1643,235 @@ namespace Sass {
...
+      // Set the first element as the `result`, keep appending to as we go down the parsedSelector vector.
+      std::vector<Selector_List*>::iterator itr = parsedSelectors.begin();
+      Selector_List* result = *itr;
+      ++itr;
+      
+      for(;itr != parsedSelectors.end(); ++itr) {
+        Selector_List* child = *itr;
+        vector<Complex_Selector*> newElements;
+
+        // For every COMPLEX_SELECTOR in `child`
```
Comment (xzyfer) [inline functions.cpp]: please remove

File: functions.cpp [hunk 10]
```diff
@@ -1566,6 +1643,235 @@ namespace Sass {
...
+      // Set the first element as the `result`, keep appending to as we go down the parsedSelector vector.
+      std::vector<Selector_List*>::iterator itr = parsedSelectors.begin();
+      Selector_List* result = *itr;
+      ++itr;
+      
+      for(;itr != parsedSelectors.end(); ++itr) {
+        Selector_List* child = *itr;
+        vector<Complex_Selector*> newElements;
+        
+        // For every COMPLEX_SELECTOR in `result`
```
Comment (xzyfer) [inline functions.cpp]: please remove

File: functions.cpp [hunk 11]
```diff
@@ -1566,6 +1643,235 @@ namespace Sass {
...
+            if(pType && pType->name() == "*") {
+              string msg("Can't append  `");
+              msg += childSeq->perform(&to_string);
+              msg += "` to `";
+              msg += parentSeqClone->perform(&to_string);;
+              msg += "`";
+              error(msg, pstate, backtrace);
+            }
+            
+            // MG TODO: Add check for namespace stuff
```
Comment (xzyfer) [inline functions.cpp]: is this still todo? open an issue?

File: subset_map.hpp [hunk 12]
```diff
@@ -91,6 +91,15 @@ namespace Sass {
     if (s.empty()) throw "internal error: subset map keys may not be empty";
     size_t index = values_.size();
     values_.push_back(value);
+    
+#ifdef DEBUG
```
Comment (xzyfer) [inline subset_map.hpp]: debug code?

File: subset_map.hpp [hunk 13]
```diff
@@ -104,6 +113,13 @@ namespace Sass {
   template<typename K, typename V>
   vector<pair<V, vector<K> > > Subset_Map<K, V>::get_kv(const vector<K>& s)
   {
+#ifdef DEBUG
```
Comment (xzyfer) [inline subset_map.hpp]: debug code?

**Comment by xzyfer** [comment]
@onedayitwillmake I've left some nitpick feedback for now. This will be a closer look when we're ready to start working on 3.3. Thanks heaps for all your work here, it's really appreciated.

One suggestion I have is to try to make the diff as clean as possible so we can focus on the changes that are relevant i.e. avoid unnecessary white changes, remove debug or commented out code

File: functions.cpp (hunk 11 above)
Comment (onedayitwillmake) [inline functions.cpp]: It's still a todo (an extra check ruby sass does), I figured i would open the issue once this is in the actual code base.

File: expand.cpp (hunk 6 above)
Comment (onedayitwillmake) [inline expand.cpp]: This one is not my own, it's from the code base, but is commented out. I re-comment it out

File: extend.cpp (hunk 7 above)
Comment (onedayitwillmake) [inline extend.cpp]: Not my comment

File: functions.cpp (hunk 9 above)
Comment (onedayitwillmake) [inline functions.cpp]: Remove the whole function?

File: functions.cpp (hunk 10 above)
Comment (onedayitwillmake) [inline functions.cpp]: Which portion?

File: extend.cpp (hunk 7 above)
Comment (xzyfer) [inline extend.cpp]: Has code been copy-pasted or moved? 
Why is it in this diff if it's not yours?

File: functions.cpp (hunk 9 above)
Comment (xzyfer) [inline functions.cpp]: This superfluous comment. Unless you think it offers some value, I doesn't to me.

File: functions.cpp (hunk 10 above)
Comment (xzyfer) [inline functions.cpp]: This superfluous comment. Unless you think it offers some value, I doesn't to me.

**Comment by xzyfer** [comment]
Superseded https://github.com/sass/libsass/pull/1261
